server/src/main/java/org/elasticsearch/indices/IndicesWriteLoadStore.java

```diff
@@ -0,0 +1,201 @@
+        }
+    }
+
+    private static String loadComposableIndexTemplate() {
+        try {
+            return Streams.readFully(IndicesWriteLoadStore.class.getResourceAsStream("/indices-write-load-distribution.json"))
+                .utf8ToString();
+        } catch (IOException e) {
+            throw new IllegalStateException("Unable to load write load distribution index template", e);
+        }
+    }
+
+    private static final Logger logger = LogManager.getLogger(IndicesWriteLoadStore.class);
+    private final BulkProcessor bulkProcessor;
+    private volatile boolean enabled;
+
+    public IndicesWriteLoadStore(ThreadPool threadPool, Client client, ClusterSettings clusterSettings, Settings settings) {
+        this(createBulkProcessor(settings, threadPool, client), clusterSettings, settings);
+    }
+
+    IndicesWriteLoadStore(BulkProcessor bulkProcessor, ClusterSettings clusterSettings, Settings settings) {
+        this.bulkProcessor = bulkProcessor;
+
+        this.enabled = ENABLED_SETTING.get(settings);
+        clusterSettings.addSettingsUpdateConsumer(ENABLED_SETTING, this::setEnabled);
+    }
+
+    static BulkProcessor createBulkProcessor(Settings settings, ThreadPool threadPool, Client client) {
+        return BulkProcessor.builder(client, new BulkProcessor.Listener() {
+            @Override
+            public void beforeBulk(long executionId, BulkRequest request) {}
+
+            @Override
+            public void afterBulk(long executionId, BulkRequest request, BulkResponse response) {
+                if (response.hasFailures()) {
+                    Map<String, String> failures = Arrays.stream(response.getItems())
+                        .filter(BulkItemResponse::isFailed)
+                        .collect(
+                            Collectors.toMap(
+                                BulkItemResponse::getId,
+                                BulkItemResponse::getFailureMessage,
+                                (msg1, msg2) -> Objects.equals(msg1, msg2) ? msg1 : msg1 + "," + msg2
+                            )
+                        );
+                    logger.warn("Failed to index some indices write load distributions: [{}]", failures);
+                }
+            }
+
+            @Override
+            public void afterBulk(long executionId, BulkRequest request, Throwable failure) {
+                long items = request.numberOfActions();
+                logger.warn(new ParameterizedMessage("Failed to index {} items into indices write load index", items), failure);
+            }
+        }, threadPool, threadPool, () -> {})
+            .setBulkActions(MAX_DOCUMENTS_PER_BULK_SETTING.get(settings))
+            .setBulkSize(MAX_BULK_SIZE_SETTING.get(settings))
+            .setConcurrentRequests(MAX_CONCURRENT_REQUESTS_SETTING.get(settings))
+            .setBackoffPolicy(BackoffPolicy.exponentialBackoff(TimeValue.timeValueMillis(1000), MAX_RETRIES_SETTING.get(settings)))
```

Comment:
I was wondering if we should just drop the indexing ops when the cluster is overloaded, we should have enough information at that point to work with the past data more or less reliably. It would be nice to have some kind of QoS so operations like this one could have lower priority than operations issued by users.